recipe-server/client/actions/show-heartbeat/index.js

```diff
@@ -16,35 +19,218 @@ export default class ShowHeartbeatAction extends Action {
+    return typeof lastShown !== 'undefined' ?
+      parseFloat(lastShown) : null;
+  }
+
+  /**
    * Checks when this survey was last shown,
    * and returns a boolean indicating if the
-   * user has seen this survey already or not.
+   * user has ever seen this survey or not.
+   *
+   * @async
+   * @return {Boolean}  Has the survey ever been shown?
    */
-  async surveyHasShown() {
-    const lastShown = await this.storage.getItem('lastShown');
+  async hasShownBefore() {
+    const lastShown = await this.getLastShown();
     // If no survey has been shown, lastShown will be falsey.
     return !!lastShown;
   }
 
+  /**
+   * Determines if this heartbeat was shown
+   * at least x days ago.
+   *
+   * @param  {Number}  days Days ago to check
+   * @return {boolean}      Has prompt been shown by that date?
+   */
+  async shownAtleastDaysAgo(days) {
+    const hasShown = await this.hasShownBefore();
+
+    if (!hasShown) {
+      return false;
+    }
+
+    // get timestamp of last shown
+    const timeLastShown = await this.getLastShown();
+
+    // get the difference between now and then
+    const timeElapsed = Date.now() - timeLastShown;
+
+    // time limit is the number of days passed in
+    // converted into milliseconds
+    const timeLimit = ONE_DAY * days;
+
+    // if the diff is smaller than the limit,
+    // that means that the last time the user saw the prompt
+    // was less than the `days` passed in
+    return timeElapsed < timeLimit;
+  }
+
+  /**
+   * Simple function to read the lastInteraction
+   * timestamp (if any) from local storage.
+   * @return {number}   Timestamp of last prompt interaction (if any)
+   */
+  async getLastInteraction() {
+    const lastInteraction = await this.storage.getItem('lastInteraction');
+
+    return typeof lastInteraction !== 'undefined' ?
+      parseFloat(lastInteraction) : null;
+  }
+
+  /**
+   * Gets the timestamp of the last prompt interaction,
+   * and returns the time (in ms) since then.
+   *
+   * @async
+   * @return {number}
+   */
+  async sinceLastInteraction() {
+    const lastInteraction = await this.getLastInteraction();
+
+    return typeof lastInteraction !== 'undefined' ?
+      Date.now() - lastInteraction : null;
+  }
+
+  /**
+   * Checks when the survey prompt last had
+   * interaction from the user (if ever),
+   * and returns a boolean indicating if the
+   * user has ever had interaction
+   *
+   * @async
+   * @return {Boolean}  Has the survey ever had interaction?
+   */
+  async hasHadInteraction() {
+    const lastInteraction = await this.getLastInteraction();
+    return !!lastInteraction;
+  }
+
+  /**
+   * Checks the repeat argument for this recipe,
+   * then determines if the recipe can be qualified as 'ran'.
+   * This ultimately decides if the prompt is shown at all
+   * to the end user.
+   *
+   * @return {boolean}        Has the heartbeat been shown?
+   */
+  async heartbeatHasRan() {
```

Comment:
Sorry, I was a bit confused here. It isn't clear to me what "run" means in this context. I was expecting that "run" meant either (1) the action has ever been executed by a runner for this recipe, or if (2) a heartbeat prompt has ever been shown for this recipe.

I think that you are describing run as (3) this recipe has fulfilled all of it's criteria for never being executed again.

The comment "Has the heartbeat ever been shown" makes me think (2). Overall I find this flow pretty hard to follow, since it is split up among a lot of different functions, and there are a lot of ill defined concepts (like "run").

All that is to say: I think changing the comment in the `@return` section will be enough to ward off enough future confusion here.